### ISSUE TITLE
Getting ValueError: Unknown loss function: cosine_proximity. Please ensure this object is passed to the `custom_objects` argument. See https://www.tensorflow.org/guide/keras/save_and_serialize#registering_the_custom_object for details. when trying to run

### ISSUE THREAD

**Issue by Waffled-II** [teticio/Deej-AI#35]
Not really sure what's up, I tried to run the code with the requirements (it couldn't find the right version of tensorflow so I just installed the latest one) and then this happens when I try to scan my music library.

C:\Users\Panda\Documents\Codestuffs\Deej-AI-master>python MP3ToVec.py Pickles mp3tovec --scan "E:\Panda\Music\iTunes\iTunes Media\Music"
2022-11-14 23:03:32.790183: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'cudart64_110.dll'; dlerror: cudart64_110.dll not found
2022-11-14 23:03:32.790836: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
Creating Track2Vec matrices
2022-11-14 23:03:35.199784: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library nvcuda.dll
2022-11-14 23:03:35.219181: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1733] Found device 0 with properties:
pciBusID: 0000:01:00.0 name: NVIDIA GeForce RTX 3080 computeCapability: 8.6
coreClock: 1.785GHz coreCount: 70 deviceMemorySize: 12.00GiB deviceMemoryBandwidth: 849.46GiB/s
2022-11-14 23:03:35.219800: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'cudart64_110.dll'; dlerror: cudart64_110.dll not found
2022-11-14 23:03:35.220358: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'cublas64_11.dll'; dlerror: cublas64_11.dll not found
2022-11-14 23:03:35.220845: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'cublasLt64_11.dll'; dlerror: cublasLt64_11.dll not found
2022-11-14 23:03:35.221412: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'cufft64_10.dll'; dlerror: cufft64_10.dll not found
2022-11-14 23:03:35.222379: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'curand64_10.dll'; dlerror: curand64_10.dll not found
2022-11-14 23:03:35.222909: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'cusolver64_11.dll'; dlerror: cusolver64_11.dll not found
2022-11-14 23:03:35.223424: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'cusparse64_11.dll'; dlerror: cusparse64_11.dll not found
2022-11-14 23:03:35.223939: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'cudnn64_8.dll'; dlerror: cudnn64_8.dll not found
2022-11-14 23:03:35.224073: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1766] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...
2022-11-14 23:03:35.224545: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX AVX2
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2022-11-14 23:03:35.225276: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1258] Device interconnect StreamExecutor with strength 1 edge matrix:
2022-11-14 23:03:35.225335: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1264]
Traceback (most recent call last):
  File "C:\Users\Panda\Documents\Codestuffs\Deej-AI-master\MP3ToVec.py", line 42, in <module>
    model = load_model(model_file)
  File "C:\Users\Panda\AppData\Local\Programs\Python\Python39\lib\site-packages\keras\saving\save.py", line 201, in load_model
    return hdf5_format.load_model_from_hdf5(filepath, custom_objects,
  File "C:\Users\Panda\AppData\Local\Programs\Python\Python39\lib\site-packages\keras\saving\hdf5_format.py", line 198, in load_model_from_hdf5
    model.compile(**saving_utils.compile_args_from_training_config(
  File "C:\Users\Panda\AppData\Local\Programs\Python\Python39\lib\site-packages\keras\saving\saving_utils.py", line 212, in compile_args_from_training_config
    loss = _deserialize_nested_config(losses.deserialize, loss_config)
  File "C:\Users\Panda\AppData\Local\Programs\Python\Python39\lib\site-packages\keras\saving\saving_utils.py", line 253, in _deserialize_nested_config
    return deserialize_fn(config)
  File "C:\Users\Panda\AppData\Local\Programs\Python\Python39\lib\site-packages\keras\losses.py", line 2020, in deserialize
    return deserialize_keras_object(
  File "C:\Users\Panda\AppData\Local\Programs\Python\Python39\lib\site-packages\keras\utils\generic_utils.py", line 698, in deserialize_keras_object
    raise ValueError(
ValueError: Unknown loss function: cosine_proximity. Please ensure this object is passed to the `custom_objects` argument. See https://www.tensorflow.org/guide/keras/save_and_serialize#registering_the_custom_object for details.

**Comment by o-aleoli** [comment]
I'm having the same problem. The culprit seems to be the changes that TF team made to model properties as it is saved. Maybe it will be necessary to refactor the code to become compatible with TF 2.10...

**Comment by teticio** [comment]
Hi there

A workaround for the timebeing is to do something along these lines when you need to load the model:

```
            self.model = load_model(
                os.path.join('model', 'speccy_model'),
                custom_objects={
                    'cosine_proximity':
                    tf.compat.v1.keras.losses.cosine_proximity
                })
```

**Comment by teticio** [comment]
I fixed the repo to work with python > 3.8 and therefore recent versions of tensorflow.
